inference-engine/src/cldnn_engine/ops/matmul.cpp

```diff
@@ -0,0 +1,247 @@
+    ngraph::Shape shape_a_aligned(shape_a), shape_b_aligned(shape_b);
+    size_t max_size = std::max(shape_a_aligned.size(), shape_b_aligned.size());
+    for (size_t i = 0, cnt = max_size - shape_a_aligned.size(); i < cnt; ++i)
+        shape_a_aligned.insert(shape_a_aligned.begin(), 1);
+    for (size_t i = 0, cnt = max_size - shape_b_aligned.size(); i < cnt; ++i)
+        shape_b_aligned.insert(shape_b_aligned.begin(), 1);
+
+    if (matmul->get_transpose_a()) {
+        std::swap(*(shape_a_aligned.end() - 1), *(shape_a_aligned.end() - 2));
+    }
+    if (matmul->get_transpose_b()) {
+        std::swap(*(shape_b_aligned.end() - 1), *(shape_b_aligned.end() - 2));
+    }
+
+    for (size_t i = 0; i < max_size - 2; ++i) {
+        if (shape_a_aligned[i] != shape_b_aligned[i] && shape_a_aligned[i] > 1 && shape_b_aligned[i] > 1) {
+            THROW_IE_EXCEPTION << "Shapes can't be aligned: " << shape_a_aligned << " " << shape_b_aligned;
+        }
+        size_t max_value = std::max(shape_a_aligned[i], shape_b_aligned[i]);
+        shape_a_aligned[i] = shape_b_aligned[i] = max_value;
+    }
+
+    return {shape_a_aligned, shape_b_aligned};
+}
+
+void CreateMatMulOp(Program& p, const std::shared_ptr<ngraph::op::v0::MatMul>& op) {
+    p.ValidateInputs(op, {2});
+    auto inputPrimitives = p.GetInputPrimitiveIDs(op);
+    std::string layerName = layer_type_name_ID(op);
+
+    auto shape_a = op->get_input_shape(0);
+    auto shape_b = op->get_input_shape(1);
+
+    bool is_fc = ngraph::is_type<ngraph::op::v0::Constant>(op->get_input_node_shared_ptr(1)) ||
+                 ngraph::is_type<ngraph::op::v0::FakeQuantize>(op->get_input_node_shared_ptr(1));
+    is_fc &= std::count_if(shape_b.begin(), shape_b.end(), [](size_t x) { return x != 1; }) <= 2;
+
+    if (is_fc) {
+        ngraph::Shape shape_a_aligned, shape_b_aligned;
+        std::tie(shape_a_aligned, shape_b_aligned) = get_aligned_shapes(shape_a, shape_b, op);
+        if (shape_a_aligned.size() < 2 || shape_b_aligned.size() < 2) {
+            THROW_IE_EXCEPTION << "MatMul " << op->get_friendly_name() << " shapes are inconsistent.";
+        }
+        size_t K = *(shape_a_aligned.end() - 1);
+        size_t O = *(shape_b_aligned.end() - 1);
+
+        auto inputName = inputPrimitives[0];
+        auto weightsName = inputPrimitives[1];
+        // Weights normalization
+        if (!op->get_transpose_b()) {
+            ngraph::Shape output_shape = shape_b;
+            std::vector<uint16_t> transpose_order(output_shape.size());
+            std::iota(transpose_order.begin(), transpose_order.end(), 0);
+            std::swap(*(transpose_order.end() - 1), *(transpose_order.end() - 2));
+
+            for (auto o = transpose_order.size(); o < 4; o++)
+                transpose_order.push_back((uint16_t)o);
+
+            auto permuteName = op->get_friendly_name() + "/transpose_b";
```

Comment:
Do we need this permute? Can we use the constant folding for transpose?

Reply:
Yeah, we need this permute to have correct FC semantic as it doesn't support transpose flags. But this primitive is folded inside the plugin.